More thorough support for retrieving OpenAPI routes in a case-sensitive manner

sander1095 opened the pull request:
# More thorough support for retrieving OpenAPI routes in a case-sensitive manner

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

This PR fixes some oversights from #59199 around retrieving OpenAPI documents case-insensitively.

## Description

This PR modifies missed files from #59199 that use the registered document name as a key for retrieving services or options. This needs to be done case-insensitively to:

- Ensure documents can be retrieved correctly
  - Trying to retrieve services around OpenAPI documents that are registered with a differently cased document name, will result in an exception, which is not desired!
- Ensure documents are generated and retrieved with the correct OpenAPI version
  - If the configuration options are retrieved with an incorrectly cased document name argument, a default instance of `OpenApiOptions` is returned, which might contain different settings than what the user had configured.

This PR also adds some more comments and improves existing ones based on the previous PR around case-insensitive OpenAPI document names.

Fixes #59175

Commit:
When retrieving optione, use the lowercase document name, too

martincostello commented on src/OpenApi/src/Extensions/OpenApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -50,7 +50,7 @@ public static IEndpointConventionBuilder MapOpenApi(this IEndpointRouteBuilder e
                 {
                     var document = await documentService.GetOpenApiDocumentAsync(context.RequestServices, context.RequestAborted);
-                    var documentOptions = options.Get(documentName);
+                    var documentOptions = options.Get(lowercasedDocumentName);
```
Is there a test that can updated that would fail before your change and pass afterwards?

sander1095 replied:
I can't really find a way to write a test for this. Looking at the source code, if the options can't find a value, it will add it instead. As this is an integration test, I don't have access to the instances and can't mock or verify method calls.

martincostello replied:
Would an HTTP request to a test server hosting an OpenAPI endpoint find the bug you're fixing?

sander1095 replied:
Never mind, I found a way! Working on it now..

sander1095 replied:
@martincostello I've added the test. Feel free to review. For now, I'll resolve this conversation as the test has been added.

sander1095 committed:
Added tests for checking if options monitor uses case-insensitive document name for OpenAPI version retrieval

captainsafia reviewed (changes requested):
Thanks for catching this and following up with a PR!

This has got me thinking that we should audit other places where we pass in documentName in the same way. As a result, it looks like we'll also want to make a change in `OpenApiDocumentProvider`. The same test around the API version should work here.

sander1095 committed:
Add XML comments to prevent bugs in the future

sander1095 committed:
Improve comments on existing test

sander1095 committed:
Add new tests for DocumentProvider which should also work case-insensitively

sander1095 committed:
Fix build by adding necessary XML comment

sander1095 committed:
Fix build of tests

sander1095 committed:
Improve comments

sander1095 committed:
Fix incorrect tests

sander1095 committed:
Fix bugs around case-sensitive document names

sander1095 commented:
@captainsafia I've implemented your requested changes, added extra comments, improved existing ones and updated the title and description of PR to better reflect the changes.

Let me know if you want me to do anything else! :)

captainsafia commented on src/OpenApi/src/Extensions/OpenApiServiceCollectionExtensions.cs:
```diff
@@ -106,6 +105,9 @@ public static IServiceCollection AddOpenApi(this IServiceCollection services, Ac
     public static IServiceCollection AddOpenApi(this IServiceCollection services)
         => services.AddOpenApi(OpenApiConstants.DefaultDocumentName);
 
+    /// <param name="services">The <see cref="IServiceCollection"/> to register services onto.</param>
```
We can remove the docstrings on private methods.

sander1095 replied:
Done

captainsafia commented on src/OpenApi/src/Services/OpenApiDocumentProvider.cs:
```diff
@@ -25,10 +25,17 @@ internal sealed class OpenApiDocumentProvider(IServiceProvider serviceProvider)
     /// <param name="writer">A text writer associated with the document to write to.</param>
     public async Task GenerateAsync(string documentName, TextWriter writer)
     {
+        // We need to retrieve the document name in a case-insensitive manner to support case-insensitive document name resolution.
```
I think we can just reference the comment in the `IServiceCollection` extensions file instead of duplicating the content here.

sander1095 replied:
Done

captainsafia commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Services/OpenApiDocumentProviderTests.cs:
```diff
@@ -30,6 +31,56 @@ public async Task GenerateAsync_ReturnsDocument()
+        // It's registered as "casesensitive" but we're passing in "CaseSensitive", which doesn't exist.
+        // Therefore, if the test doesn't throw, we know it has passed correctly.
+        // We still do a small check to validate the document, just in case. But the main test is that it doesn't throw.
+        ValidateOpenApiDocument(stringWriter, _ => { });
```
Nit: Add an assertion here to validate the OpenAPI version as above?

sander1095 replied:
Done

sander1095 commented:
Thanks for the review! I'll implement requested changes after my holiday (in ~2 weeks)

sander1095 committed:
Implemented PR comments

sander1095 commented:
@captainsafia I've processed the PR comments!

captainsafia commented:
/azp run
